docs/upgrade.html

```diff
@@ -45,6 +45,22 @@ <h5><a id="upgrade_240_notable" href="#upgrade_240_notable">Notable changes in 2
         The old overloaded functions are deprecated and we would recommend users to make their code changes to leverage the new methods (details
         can be found in <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-520%3A+Add+overloaded+Consumer%23committed+for+batching+partitions">KIP-520</a>).
     </li>
+    <li>We are introducing incremental cooperative rebalancing to the clients' group protocol, which allows consumers to keep all of their assigned partitions during a rebalance
+        and at the end revoke only those which must be migrated to another consumer for overall cluster balance. To upgrade from eager to cooperative rebalancing, you must set all your clients
+        to use the new <code>CooperativeStickyAssignor</code>. This can be done with two rolling bounces: during the first one, add "cooperative-sticky" to the list of supported assignors
+        for each member (without removing the previous assignor -- note that if previously using the default, you must include that explicitly as well) before you bounce and/or upgrade it.
+        Once the entire group is on 2.4+ and all members have the <code>CooperativeStickyAssignor</code> among their supported assignors, perform a second rolling bounce and remove the
+        other assignor(s) so that by the end all members support only the cooperative protocol. For further details on the safe upgrade path or the cooperative rebalancing protocol,
+        see <a href="https://cwiki.apache.org/confluence/x/vAclBg">KIP-429</a>.
+    </li>
+    <li>There are some behavioral changes to the <code>ConsumerRebalanceListener</code>, as well as a new API. The <code>onPartitionsLost</code> method has been added to allow
```

Comment:
This is a meta comment: I suggest we re-structure the explanation a bit:

1. The assignor API changes with owned partitions (including minor renamings), with new built-in cooperative assignor.
1.1. How to write your own cooperative assignor (it's your responsibility to not move assignment immediately).
1.2. Upgrade to cooperative assignors with two rolling bounces.
2. Rebalance listener changes, including new APIs, and the triggering semantics.
2.1. Rebalance listener exception thrown handling.

Reply:
Refactored this paragraph a bit differently, let me know how it looks now